File: apps/api/app/schemas/root_cause.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from ..models import (
    RootCauseAnalysisStatus, RootCauseFindingConfidence, RootCauseFindingSeverity,
    RootCauseFindingStatus,
)


class RootCauseAnalysisCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    diagnostic_run_id: UUID


class RootCauseFindingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    analysis_id: UUID
    rule_id: str
    severity: RootCauseFindingSeverity
    title: str
    summary: str
    organization_id: UUID
    device_id: UUID
    diagnostic_result_id: UUID
    category: str
    status: RootCauseFindingStatus
    confidence: RootCauseFindingConfidence
    explanation: str
    evidence: dict
    created_at: datetime
    updated_at: datetime


class RootCauseAnalysisRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    organization_id: UUID
    diagnostic_run_id: UUID
    device_id: UUID
    initiated_by_user_id: UUID | None
    provider: str
    status: RootCauseAnalysisStatus
    completed_at: datetime | None
    cancelled_at: datetime | None
    error_code: str | None
    created_at: datetime
    updated_at: datetime


class RootCausePageMeta(BaseModel):
    page: int
    page_size: int
    total: int


class RootCauseAnalysisPage(BaseModel):
    items: list[RootCauseAnalysisRead]
    meta: RootCausePageMeta


class RootCauseFindingPage(BaseModel):
    items: list[RootCauseFindingRead]
    meta: RootCausePageMeta

File: apps/web/src/types/rootCause.ts
/** apps/api/app/schemas/root_cause.py::RootCauseAnalysisRead */
export type RootCauseAnalysis = {
  id: string
  organization_id: string
  diagnostic_run_id: string
  device_id: string
  initiated_by_user_id: string | null
  provider: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  completed_at: string | null
  cancelled_at: string | null
  error_code: string | null
  created_at: string
  updated_at: string
}

/** apps/api/app/schemas/root_cause.py::RootCauseFindingRead */
export type RootCauseFinding = {
  id: string
  analysis_id: string
  rule_id: string
  severity: 'low' | 'medium' | 'high' | 'critical'
  title: string
  summary: string
  organization_id: string
  device_id: string
  diagnostic_result_id: string
  category: string
  status: 'identified' | 'likely' | 'insufficient_evidence'
  confidence: 'high' | 'medium' | 'low'
  explanation: string
  evidence: Record<string, unknown>
  created_at: string
  updated_at: string
}
